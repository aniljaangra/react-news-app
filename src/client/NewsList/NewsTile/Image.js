/**
 * Created by Anil Jangra on 5/23/2021
 */
import React, { useState } from 'react';

const Image = ({ src }) => {
  const [isLoading, setIsLoading] = useState(true);

  function onLoad() {
    setIsLoading(false);
  }

  return (
    <>
      <img
        className="card-img-top"
        src="https://via.placeholder.com/300x200/fff"
        style={{ display: isLoading ? 'block' : 'none' }}
        alt="IMG"
      />
      <img
        className="card-img-top"
        src={src}
        style={{ display: isLoading ? 'none' : 'block' }}
        onLoad={onLoad}
        alt="IMG"
      />
    </>
  );
};

export default Image;
