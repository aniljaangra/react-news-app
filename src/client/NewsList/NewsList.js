import React from 'react';
import PropTypes from 'prop-types';
import NewsTile from './NewsTile/NewsTile';

function NewsList({
  articles, error, loading, query,
}) {
  const noArticles = articles.length === 0;

  const renderLoadingOrError = () => {
    if (loading && noArticles) {
      return <h6 className="mt-5">Loadingâ€¦</h6>;
    } if (error) {
      return <h6 className="mt-5 alert">Failed to load the News.</h6>;
    }
    return null;
  };

  const renderQueryString = () => {
    if (query) {
      return (
        <h6 className="mt-5">
          Showing search results for
          {' '}
          {`"${query}"`}
        </h6>
      );
    }
    return null;
  };

  const renderArticles = () => (!noArticles) && (
  <div className="row py-3 mt-3">
    {articles.map((news) => <NewsTile news={news} key={news.url} />)}
    {!noArticles && loading && (
    <div className="spinner-grow text-info m-5" role="status" />
    )}
  </div>
  );

  return (
    <div className="container-fluid min-vh-100">
      <div className="container mt-6">
        {renderQueryString()}
        {renderLoadingOrError()}
        {renderArticles()}
      </div>
    </div>
  );
}

export default NewsList;

NewsList.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    urlToImage: PropTypes.string.isRequired,
    source: PropTypes.shape({
      name: PropTypes.string.isRequired,
    }).isRequired,
  })).isRequired,
  error: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  query: PropTypes.string.isRequired,
};
