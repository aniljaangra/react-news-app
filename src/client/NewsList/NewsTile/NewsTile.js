import React from 'react';
import Image from './Image';

function NewsTile({ news }) {
  return (
    <div className="col-lg-4 col-md-6 col-sm-12 d-flex align-items-stretch">
      <div
        className="card shadow mb-5 p-2 bg-white rounded"
      >
        <Image src={news.urlToImage} />
        <div className="card-body">
          <h5 className="card-title">
            {news.title}
          </h5>
          <p className="card-text">
            {news.description}
          </p>
        </div>
        <a
          href={news.url}
          className="btn btn-primary"
          target="_blank"
          style={{ width: '50%', margin: '5% auto 10% auto' }}
          rel="noreferrer"
        >
          {news.source.name}
        </a>
      </div>
    </div>
  );
}
export default NewsTile;
