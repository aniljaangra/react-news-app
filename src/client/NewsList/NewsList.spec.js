/**
 * Created by Anil Jangra on 5/24/2021
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import NewsList from './NewsList';
import NewsTile from './NewsTile/NewsTile';

describe('NewsTile', () => {
  const articles = [{
    title: 'sample news',
    description: 'sample description',
    url: 'https://localhost',
    urlToImage: 'http://localhost/image',
    source: {
      name: 'sample source',
    },
  }];
  it('renders NewsList component correctly', () => {
    const tree = shallow(<NewsList articles={articles} error="" query="" loading={false} />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('renders loading when initial set of articles are loading', () => {
    const wrapper = shallow(<NewsList articles={[]} error="" query="" loading />);
    expect(wrapper.text()).toContain('Loading');
  });

  it('renders NewsTile when initial set of articles is loaded', () => {
    const wrapper = shallow(<NewsList articles={articles} error="" query="" loading={false} />);
    expect(wrapper.find(NewsTile).length).toEqual(1);
  });

  it('should show error when there is error loading API data', () => {
    const wrapper = shallow(<NewsList articles={[]} error="error" query="" loading={false} />);
    expect(wrapper.text()).toEqual('Failed to load the News.');
  });

  it('should show query when there is query based search', () => {
    const wrapper = shallow(<NewsList articles={[]} error="" query="search query" loading={false} />);
    expect(wrapper.text()).toContain('search query');
  });
});
