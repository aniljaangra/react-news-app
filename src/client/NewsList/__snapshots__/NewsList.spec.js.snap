// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsTile renders NewsList component correctly 1`] = `
<div
  className="container-fluid min-vh-100"
>
  <div
    className="container mt-6"
  >
    <div
      className="row py-3 mt-3"
    >
      <NewsTile
        key="https://localhost"
        news={
          Object {
            "description": "sample description",
            "source": Object {
              "name": "sample source",
            },
            "title": "sample news",
            "url": "https://localhost",
            "urlToImage": "http://localhost/image",
          }
        }
      />
    </div>
  </div>
</div>
`;
